atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/TransactionScopedCacheImpl.java

```diff
@@ -129,15 +184,31 @@ private void ensureNotFinalised() {
     }
 
     private synchronized Map<Cell, byte[]> processRemoteRead(
-            TableReference tableReference, CacheLookupResult cacheLookup, Map<Cell, byte[]> remoteReadValues) {
+            TableReference tableReference,
+            Map<Cell, CacheValue> cacheHits,
+            Set<Cell> cacheMisses,
+            Map<Cell, byte[]> remoteReadValues) {
         valueStore.cacheRemoteReads(tableReference, remoteReadValues);
-        cacheEmptyReads(tableReference, cacheLookup.missedCells(), remoteReadValues);
+        cacheEmptyReads(tableReference, cacheMisses, remoteReadValues);
         return ImmutableMap.<Cell, byte[]>builder()
                 .putAll(remoteReadValues)
-                .putAll(filterEmptyValues(cacheLookup.cacheHits()))
+                .putAll(filterEmptyValues(cacheHits))
                 .build();
     }
 
+    private synchronized NavigableMap<byte[], RowResult<byte[]>> processRemoteReadRows(
+            TableReference tableReference, Set<Cell> cacheMisses, Map<byte[], RowResult<byte[]>> remoteReadValues) {
+        Map<Cell, byte[]> rowsReadAsCells = remoteReadValues.values().stream()
+                .map(RowResult::getCells)
```

Comment:
nit: a utility method here might look good

Reply:
I tried, but it's messy